packages/s3-store/index.ts

```diff
@@ -221,11 +217,40 @@ export class S3Store extends DataStore {
 
   private async prependIncompletePart(
     newChunkPath: string,
-    previousIncompletePart: Uint8Array
-  ): Promise<void> {
-    const newChunk = await fsProm.readFile(newChunkPath)
-    const combined = Buffer.concat([previousIncompletePart, newChunk])
-    await fsProm.writeFile(newChunkPath, combined)
+    previousIncompletePart: Readable
+  ): Promise<number> {
+    const tempPath = `${newChunkPath}-prepend`
+    try {
+      let incompletePartSize = 0
+
+      const byteCounterTransform = new stream.Transform({
+        transform(chunk, _, callback) {
+          incompletePartSize += chunk.length
+          callback(null, chunk)
+        },
+      })
+
+      // write to temporary file, truncating if needed
+      await streamProm.pipeline(
+        previousIncompletePart,
+        byteCounterTransform,
+        fs.createWriteStream(tempPath)
+      )
+      // append to temporary file
+      await streamProm.pipeline(
+        fs.createReadStream(newChunkPath),
+        fs.createWriteStream(tempPath, {flags: 'a'})
+      )
```

Comment:
Would this work too?

```suggestion
    await streamProm.pipeline(
      previousIncompletePart,
      byteCounterTransform,
      // Open file for appending. The file is created if it does not exist.
      fs.createWriteStream(tempPath, { flags: 'a' }),
      fs.createReadStream(newChunkPath),
      fs.createWriteStream(tempPath, { flags: 'a' })
    );
```

Reply:
No. Pipeline starts with readable stream, following zero or more transform streams and end with writable stream. What we need is to concatenate two streams.

I'm opening a file two times because it is less error prone to use `stream.pipeline(...)`. When pipeline is finished it closes write stream so it can not be reused for second pipeline. More efficient approach would be to manually `pipe` from both streams and manually handle errors and closing at the end.